File: 20230101_Sunday/20230101.go
// Leetcode: 290. Word Pattern.
// https://leetcode.com/problems/word-pattern/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 0 ms, faster than 100.00% of Go online submissions for Word Pattern.
// Memory Usage: 1.9 MB, less than 65.81% of Go online submissions for Word Pattern.
// 2023.01.01 Daily Challenge.
import "strings"
import "fmt"
var p = fmt.Println
func wordPattern(pattern string, s string) bool {
  wrds := strings.Fields(s)
  if len(pattern) != len(wrds) { return false }
  hChars := make(map[rune]string)
  hUsed := make(map[string]bool)
  for i,char := range pattern {
    word := wrds[i]
    value, has_key := hChars[char]
    if  has_key {
      if word != value { return false }
    } else {
      // Check if the word is used.
      _, used := hUsed[word]
      if used { return false }
      hUsed[word] = true
      hChars[char] = word
    }
  }
  return true
}

File: 20230101_Sunday/20230101.rb
# Leetcode: 290. Word Pattern.
# https://leetcode.com/problems/word-pattern/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 107 ms, faster than 20.00% of Ruby online submissions for Word Pattern.
# Memory Usage: 211 MB, less than 87.50% of Ruby online submissions for Word Pattern.
# 2023.01.01 Daily challenge.
# @param {String} pattern
# @param {String} s
# @return {Boolean}
def word_pattern(pattern, s)
  h = {}
  h1 = {}
  i = 0 # char poiner.
  s.split(' ').each do |w|
    return false if i >= pattern.size
    key = pattern[i]
    value = h[key]
    used = h1[w]
    if value
      return false if w != value
    else
      return false if used 
      h[key] = w
      h1[w] = true
    end
    i += 1
  end
  return i == pattern.size
end
